src/frollR.c

```diff
@@ -21,83 +22,82 @@ SEXP coerceToRealListR(SEXP obj) {
   UNPROTECT(protecti);
   return x;
 }
-
-SEXP frollfunR(SEXP fun, SEXP obj, SEXP k, SEXP fill, SEXP algo, SEXP align, SEXP narm, SEXP hasnf, SEXP adaptive) {
+// validate and coerce to integer or list of integer
+SEXP coerceK(SEXP obj, bool adaptive) {
   int protecti = 0;
-  const bool verbose = GetVerbose();
-
-  if (!xlength(obj))
-    return(obj);                                                // empty input: NULL, list()
-  double tic = 0;
-  if (verbose)
-    tic = omp_get_wtime();
-  SEXP x = PROTECT(coerceToRealListR(obj)); protecti++;
-  R_len_t nx=length(x);                                         // number of columns to roll on
-
-  if (xlength(k) == 0)                                          // check that window is non zero length
-    error(_("n must be non 0 length"));
-
-  if (!IS_TRUE_OR_FALSE(adaptive))
-    error(_("%s must be TRUE or FALSE"), "adaptive");
-  bool badaptive = LOGICAL(adaptive)[0];
-
-  R_len_t nk = 0;                                               // number of rolling windows, for adaptive might be atomic to be wrapped into list, 0 for clang -Wall
-  SEXP ik = R_NilValue;                                         // holds integer window width, if doing non-adaptive roll fun
-  SEXP kl = R_NilValue;                                         // holds adaptive window width, if doing adaptive roll fun
-  if (!badaptive) {                                             // validating n input for adaptive=FALSE
-    if (isNewList(k))
+  SEXP ans = R_NilValue;
+  if (!adaptive) {
+    if (isNewList(obj))
       error(_("n must be integer, list is accepted for adaptive TRUE"));
-
-    if (isInteger(k)) {                                        // check that k is integer vector
-      ik = k;
-    } else if (isReal(k)) {                                    // if n is double then convert to integer
-      ik = PROTECT(coerceVector(k, INTSXP)); protecti++;
+    if (isInteger(obj)) {
+      ans = obj;
+    } else if (isReal(obj)) {
+      ans = PROTECT(coerceVector(obj, INTSXP)); protecti++;
     } else {
       error(_("n must be integer"));
     }
-
-    nk = length(k);
-    R_len_t i=0;                                                // check that all window values positive
-    while (i < nk && INTEGER(ik)[i] > 0) i++;
+    int nk = length(obj);
+    R_len_t i = 0;
+    int *iik = INTEGER(ans);
+    while (i < nk && iik[i] > 0) i++;
     if (i != nk)
       error(_("n must be positive integer values (> 0)"));
```

Comment:
This feels strange since `frollmean(1:2, list(c(0L, 0L)), adaptive=TRUE)` will work while ` frollmean(1:2, 0L, adaptive=FALSE)` won't

Reply:
good spot, adaptive branch was not checking for positive values